numpy/core/src/multiarray/compiled_base.c

```diff
@@ -825,29 +830,36 @@ static int sequence_to_arrays(PyObject *seq,
     }
 
     for (i = 0; i < count; ++i) {
-        PyObject *item = PySequence_GetItem(seq, i);
+        /* ensure dtype is based off dtype_guess */
+        PyArray_Descr *dtype = dtype_guess;
+        item = PySequence_GetItem(seq, i);
         if (item == NULL) {
-            while (--i >= 0) {
-                Py_DECREF(op[i]);
-                op[i] = NULL;
-            }
-            return -1;
+            goto fail;
         }
-
-        op[i] = (PyArrayObject *)PyArray_FROM_O(item);
+        Py_INCREF(dtype);
+        if (PyArray_DTypeFromObject(item, NPY_MAXDIMS, &dtype) < 0) {
+            goto fail;
+        }
+        if (!(PyDataType_ISINTEGER(dtype) || PyDataType_ISBOOL(dtype))) {
+            PyErr_SetString(PyExc_TypeError, "only int indices permitted");
```

Comment:
We should look to deprecate booleans here

Reply:
removed in a later commit